src/base/bittorrent/tracker.cpp

```diff
@@ -32,235 +33,411 @@
 {
-    if (announceReq.peer.port == 0) return;
-
     if (!m_torrents.contains(announceReq.infoHash)) {
-        // Unknown torrent
-        if (m_torrents.size() == MAX_TORRENTS) {
-            // Reached max size, remove a random torrent
+        // Reached max size, remove a random torrent
+        if (m_torrents.size() >= MAX_TORRENTS)
             m_torrents.erase(m_torrents.begin());
-        }
     }
 
-    // Register the user
-    PeerList &peers = m_torrents[announceReq.infoHash];
-    if (!peers.contains(announceReq.peer.uid())) {
-        // Unknown peer
-        if (peers.size() == MAX_PEERS_PER_TORRENT) {
-            // Too many peers, remove a random one
-            peers.erase(peers.begin());
-        }
-    }
-    peers[announceReq.peer.uid()] = announceReq.peer;
+    TorrentStats &torrentStats = m_torrents[announceReq.infoHash];
+    QSet<Peer> &peers = torrentStats.peers;
+
+    // Too many peers, remove a random one
+    if (peers.size() >= MAX_PEERS_PER_TORRENT)
+        torrentStats.removePeer(*peers.begin());
+
+    torrentStats.setPeer(announceReq.peer);
 }
 
 void Tracker::unregisterPeer(const TrackerAnnounceRequest &announceReq)
 {
-    if (announceReq.peer.port == 0) return;
+    const auto torrentStatsIter = m_torrents.find(announceReq.infoHash);
+    if (torrentStatsIter == m_torrents.end())
+        return;
 
-    if (m_torrents[announceReq.infoHash].remove(announceReq.peer.uid()) > 0)
-        qDebug("Tracker: Peer stopped downloading, deleting it from the list");
+    torrentStatsIter->removePeer(announceReq.peer);
+
+    if (torrentStatsIter->peers.isEmpty())
+        m_torrents.erase(torrentStatsIter);
 }
 
-void Tracker::replyWithPeerList(const TrackerAnnounceRequest &announceReq)
+void Tracker::prepareAnnounceResponse(const TrackerAnnounceRequest &announceReq)
 {
-    // Prepare the entry for bencoding
-    lt::entry::dictionary_type replyDict;
-    replyDict["interval"] = lt::entry(ANNOUNCE_INTERVAL);
+    const TorrentStats &torrentStats = m_torrents[announceReq.infoHash];
+
+    lt::entry::dictionary_type replyDict {
+        {ANNOUNCE_RESPONSE_INTERVAL, ANNOUNCE_INTERVAL},
+        {ANNOUNCE_RESPONSE_COMPLETE, torrentStats.seeders},
+        {ANNOUNCE_RESPONSE_INCOMPLETE, (torrentStats.peers.size() - torrentStats.seeders)},
+
+        // [BEP-24] Tracker Returns External IP
+        {ANNOUNCE_RESPONSE_EXTERNAL_IP, toBigEndianByteArray(announceReq.peer.address).toStdString()}
+    };
+
+    // peer list
+    // [BEP-7] IPv6 Tracker Extension (partial support)
+    // [BEP-23] Tracker Returns Compact Peer Lists
+    if (announceReq.compact) {
+        lt::entry::list_type peerList;
+        lt::entry::list_type peer6List;
+
+        int counter = 0;
+        for (const Peer &peer : torrentStats.peers) {
+            if (counter++ >= announceReq.numwant)
+                break;
+
+            const QHostAddress claimedAddressIP {QString(peer.claimedAddress)};
```

Comment:
IMO, redundant conversion. You convert it back to byte array below.
You've struggled with this conversions recently.

Reply:
Since user (tracker client) can put anything to claimedAddress, I need to verify it is really an IP address and use it first over `peer.address`.